astropy/table/table.py

```diff
@@ -781,7 +781,16 @@ def __init__(
                 f"__init__() got unexpected keyword argument {next(iter(kwargs.keys()))!r}"
             )
 
-        if isinstance(data, np.ndarray) and data.shape == (0,) and not data.dtype.names:
+        # Treat any empty numpy array as None, except for structured arrays since they
+        # provide column names and dtypes.
+        #
+        # Init with rows=[] or data=[] (or tuples) is allowed and taken to mean no data.
+        # This allows supplying names and dtype if desired. `data=[]` is ambiguous,
+        # because it could mean no columns, or it could mean no rows for list of dict.
+        # For compatibility with the latter, interpret data=[] as data=None.
+        if (
+            isinstance(data, np.ndarray) and data.size == 0 and not data.dtype.names
```

Comment:
`Table(rows=[], names=["a", "b"], dtype=[int, float])` is explicitly documented as accepted, so I'm not sure I get why an empty array *with* `dtype.names` wouldn't be (except maybe for sake of simplicity in the implementation, but it seems we're well passed that ?)

Reply:
The zero-length structured array is accepted and handled in later code. You don't want to set `data = None` here (empty table) because you would lose the column names from `dtype.names`.
```
In [6]: ta
Out[6]: array([], dtype=[('a', '<i8'), ('b', '<f8'), ('c', '<U1')])

In [7]: Table(ta)
Out[7]: 
<Table length=0>
  a      b     c  
int64 float64 str1
----- ------- ----

```

Reply:
oh, I see. I'd suggest to add an explicit test case for it then, next to the one already introduced.

Reply:
Of initialization from a zero length record array.

Reply:
Is this the correct behavior?

```python
>>> import numpy as np
>>> from astropy.table import Table
>>> a = np.array([], dtype=[('foo', int), ('bar', int)])
>>> Table(data=a)
<Table length=0>
 foo   bar
int64 int64
----- -----

>>> Table(rows=a)
<Table length=0>
<No columns>
```

Reply:
No. Probably in both cases the names and data types should be populated from the recarray's dtype.

Reply:
I will need @taldcroft to advise because that is the behavior with this patch as-is.

Reply:
Nothing like a good game of whack-a-mole. This seems to do the trick?
```
index 335d88cbfb..8c0857de24 100644
--- a/astropy/table/table.py
+++ b/astropy/table/table.py
@@ -746,7 +746,9 @@ class Table:
             names_from_list_of_dict = _get_names_from_list_of_dict(rows)
             if names_from_list_of_dict:
                 data = rows
-            elif isinstance(rows, self.Row):
+            elif isinstance(rows, self.Row) or (
+                isinstance(rows, np.ndarray) and rows.dtype.names
+            ):
                 data = rows
             else:
                 data = list(zip(*rows))
```